components/opensea/sources/events/events.mjs

```diff
@@ -0,0 +1,55 @@
+import { axios } from "@pipedream/platform";
+import opensea from "../../opensea.app.mjs";
+
+export default {
+  name: "New OpenSea Collection Events",
+  version: "0.0.1",
+  key: "opensea-events",
+  description: "Emit new filtered OpenSea events",
+  props: {
+    opensea,
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60,
+      },
+    },
+    contractAddress: {
+      type: "string",
+      label: "Contract Address",
+      description: "Collection contract address",
+    },
+    eventType: {
+      type: "string",
+      options: [
+        "sales",
+        "listings",
+      ],
+      label: "Event Type",
+      description: "OpenSea event type",
+    },
+  },
+  dedupe: "greatest",
```

Comment:
The `dedupe` strategy `greatest` works well with apps that use incrementing `ids` such as [GitHub](https://pipedream.com/docs/components/quickstart/nodejs/sources/#dedupe-the-events).

I'm not sure if OpenSea `event` `ids` are incremental; I couldn't find it in the docs.
If not, since the [retrieving events endpoint](https://docs.opensea.io/reference/retrieving-asset-events) accepts the parameter `occurred_after`, in order to prevent duplicate events a `db` prop may be used to cache the `timestamp` of the last emitted `event`.

For example,

```javascript
props: {
  db: "$.service.db",
},
methods: {
  getLastTimestamp() {
    return this.db.get("lastTimestamp");
  },
  setLastTimestamp(ts) {
    this.db.set("lastTimestamp", ts);
  },
},
async run() {
  ...

  let lastTimestamp = this.getLastTimestamp();
  const resp = this.opensea.retrieveEvents({
    contract,
    eventType,
    lastTimestamp,
  });

  resp.asset_events
    .filter((event) => event.created_date > lastTimestamp)
    .forEach((event) => {
      this.$emit(event, {
        id: event.id,
        ts: +new Date(event.created_date),
      });
      lastTimestamp = event.created_date;
    });
  this.setLastTimestamp(lastTimestamp);

  ...
},
```

Reply:
The event ids are incremental.